homeassistant/components/bedrock_agent/__init__.py

```diff
@@ -0,0 +1,293 @@
+            ),
+            vol.Optional(CONST_SERVICE_PARAM_FILENAMES, default=[]): vol.All(
+                cv.ensure_list, [cv.string]
+            ),
+        }
+    )
+
+    hass.services.async_register(
+        DOMAIN,
+        "cognitive_task",
+        async_cognitive_task,
+        schema=COGNITIVE_TASK_SCHEMA,
+        supports_response=SupportsResponse.ONLY,
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    conversation.async_unset_agent(hass, entry)
+    return True
+
+
+async def options_update_listener(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Handle options update."""
+    # await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+class BedrockAgent(conversation.AbstractConversationAgent):
+    """Bedrock conversation agent."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the agent."""
+        self.hass = hass
+        self.entry = entry
+        self.history: dict[str, list[dict]] = {}
+        self.bedrock = boto3.client(
+            service_name="bedrock-runtime",
+            region_name=self.entry.data[CONST_REGION],
+            aws_access_key_id=self.entry.data[CONST_KEY_ID],
+            aws_secret_access_key=self.entry.data[CONST_KEY_SECRET],
+        )
+        self.bedrock_agent = boto3.client(
+            service_name="bedrock-agent-runtime",
+            region_name=self.entry.data[CONST_REGION],
+            aws_access_key_id=self.entry.data[CONST_KEY_ID],
+            aws_secret_access_key=self.entry.data[CONST_KEY_SECRET],
+        )
+
+    @property
+    def supported_languages(self) -> list[str] | Literal["*"]:
+        """Return a list of supported languages."""
+        return MATCH_ALL
+
+    @staticmethod
+    def supported_models() -> list[str]:
+        """Return a list of supported models."""
+        return CONST_MODEL_LIST
+
+    async def async_call_bedrock(self, question) -> str:
+        """Return result from Amazon Bedrock."""
+
+        question = self.entry.options[CONST_PROMPT_CONTEXT] + question
+
+        modelId = self.entry.options[CONST_MODEL_ID]
+        knowledgebaseId = self.entry.options.get(CONST_KNOWLEDGEBASE_ID) or ""
+
+        if knowledgebaseId != "":
+            agent_input = {"text": question}
+            agent_retrieveAndGenerateConfiguration = {
+                "knowledgeBaseConfiguration": {
+                    "knowledgeBaseId": knowledgebaseId,
+                    "modelArn": modelId,
+                },
+                "type": "KNOWLEDGE_BASE",
+            }
+
+            bedrock_agent_response = await self.hass.async_add_executor_job(
+                partial(
+                    self.bedrock_agent.retrieve_and_generate,
+                    input=agent_input,
+                    retrieveAndGenerateConfiguration=agent_retrieveAndGenerateConfiguration,
```

Comment:
Does the bedrock API support history of the conversation? If yes, we should separate it with `user_input.conversation_id`